lib/std/mem.zig

```diff
@@ -876,7 +870,86 @@ pub fn indexOfPos(comptime T: type, haystack: []const T, start_index: usize, nee
     return null;
 }
 
+fn boyerMooreHorspoolPreprocessReverse(pattern: []const u8, table: *[256]usize) void {
+    for (table) |*c| {
+        c.* = pattern.len;
+    }
+
+    var i: usize = pattern.len - 1;
+    while (i > 0) : (i -= 1) {
+        table[pattern[i]] = i;
+    }
+}
+
+fn boyerMooreHorspoolPreprocess(pattern: []const u8, table: *[256]usize) void {
+    for (table) |*c| {
+        c.* = pattern.len;
+    }
+
+    var i: usize = 0;
+    while (i < pattern.len - 1) : (i += 1) {
+        table[pattern[i]] = pattern.len - 1 - i;
+    }
+}
+/// Find the index in a slice of a sub-slice, searching from the end backwards.
+/// To start looking at a different index, slice the haystack first.
+// Reverse boyer-moore-horspool algorithm
+pub fn lastIndexOf(comptime T: type, haystack: []const T, needle: []const T) ?usize {
+    if (needle.len > haystack.len) return null;
+    if (needle.len == 0) return haystack.len;
+
+    if (!meta.trait.hasUniqueRepresentation(T) or haystack.len < 32 or needle.len <= 2)
```

Comment:
How were these thresholds chosen?  On a modern CPU I would expect the linear search to be faster up to haystack sizes of at least 1000, even with a needle size of 32.  I think some perf testing is needed here.

Reply:
You are right, i haven't tested for ideal haystack size, only for the needle size. Performance testing for ideal haystack size is a bit more annoying because the compiler optimizes the code unless i use an external source with millions of words of size 0-1000

Reply:
This will probably depend a lot on whether you optimize for average or worst case input.
Malicious input has to be considered when optimizing for the average case. Worst case performance should still be acceptable.

Reply:
My general preference, which I believe is in line with most of the zig language & std lib, is to optimize for worst case input.

Reply:
Ok so I wrote some tests for this and the results are.. complicated. In general the optimal cuttoff is a needle size between 2-8, depending on how large the haystack is. But surprisingly the old version of indexOf gets faster the larger the needle. So for certain sizes of haystack the older version is faster for a needle size of 1-4 and slower for 4-16/32 and then faster for a larger number. After a certain size of haystack (somewhere between 120 and ~256) the older version is faster for a needle of size 1-4 but always slower for a larger needle. I need to create a graph of these results or something to display it properly and then that could be used for future benchmarks. It will take some time though..

Reply:
And of course these are the results for my cpu (i5 4690k), im not sure how the results would look like for other cpus with different cpu caches

Reply:
Is there any other search algorithm that is more stable? i looked in glibc and they have similar cuttoffs between different algorithms

Reply:
Here are the benchmarks: https://imgur.com/a/EH17VNo . The naive linear case is also very good, need to check llvm is doing some optimization magic

Reply:
And then here is the benchmark where indexOfPosLinear is called when haystack.len < 52 or needle.len <= 4: https://imgur.com/a/EQJRaBk

Reply:
Also here is benchmark with a 5mb haystack: https://i.imgur.com/hmd9rGt.png